eta/core/events.py

```diff
@@ -27,7 +31,194 @@
+            index: (optional) an index assigned to the event
+        '''
+        event = Event(start, stop)
+        series = EventSeries([event])
+        return DetectedEvent(label, series, confidence, index)
+
+    @classmethod
+    def from_dict(cls, d):
+        '''Constructs a DetectedEvent from a JSON dictionary.'''
+        attrs = d.get("attrs", None)
+        if attrs is not None:
+            attrs = AttributeContainer.from_dict(attrs)
+
+        return cls(
+            d["label"],
+            EventSeries.from_dict(d["event_series"]),
+            confidence=d.get("confidence", None),
+            index=d.get("index", None),
+            score=d.get("score", None),
+            attrs=attrs,
+        )
+
+
+class DetectedEventContainer(Container):
+    '''Base class for containers that store lists of `DetectedEvent`s.'''
+
+    _ELE_CLS = DetectedEvent
+    _ELE_CLS_FIELD = "_EVENT_CLS"
+    _ELE_ATTR = "events"
+
+    def get_labels(self):
+        '''Returns a set containing the labels of the DetectedEvents.'''
+        return set(event.label for event in self)
+
+    def sort_by_confidence(self, reverse=False):
+        '''Sorts the event list by confidence.
+
+        Events whose confidence is None are always put last.
+
+        Args:
+            reverse: whether to sort in descending order. The default is False
+        '''
+        self.sort_by("confidence", reverse=reverse)
+
+    def sort_by_index(self, reverse=False):
+        '''Sorts the event list by index.
+
+        Events whose index is None are always put last.
+
+        Args:
+            reverse: whether to sort in descending order. The default is False
+        '''
+        self.sort_by("index", reverse=reverse)
+
+    def sort_by_score(self, reverse=False):
+        '''Sorts the event list by score.
+
+        Events whose score is None are always put last.
+
+        Args:
+            reverse: whether to sort in descending order. The default is False
+        '''
+        self.sort_by("score", reverse=reverse)
+
+    def filter_by_schema(self, schema):
+        '''Filters the events/attributes from this container that are not
+        compliant with the given schema.
+
+        Args:
+            schema: an ImageLabelsSchema or VideoLabelsSchema
+        '''
+        filter_func = lambda event: event.label in schema.events
+        self.filter_elements([filter_func])
+        for event in self:
+            if event.has_attributes:
+                event.attrs.filter_by_schema(schema.events[event.label])
+
+    def remove_events_without_attrs(self, labels=None):
+        '''Filters the events from this container that do not have attributes.
+
+        Args:
+            labels: an optional list of DetectedEvent label strings to which
+                to restrict attention when filtering. By default, all event
+                are processed
+        '''
+        filter_func = lambda event: (
+            (labels is not None and event.label not in labels)
```

Comment:
I think this is correct as it is. The extra `labels` keyword is used to *restrict* the filter to certain labels. So objects without a label in `labels` are not targeted and therefore not filtered out.

These `lambda`s are pretty easy write, so if anything, I would say this function may not need to be a proper method for the class.

Reply:
If this is the case, then the doc string wording is a bit confusing.  I would call it a "black list" of labels to express what @benjaminpkane is saying.

Reply:
@mattphotonman it's implemented correctly and the docstring is correct. you're probably confused because you expect `filter_elements` to remove elements for which the filter is `True`, but it *keeps* elements them